File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/firmware/BaseLoadFirmwareCommand.java
package com.worldventures.dreamtrips.wallet.service.command.firmware;

import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.domain.entity.FirmwareUpdateData;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableFirmwareUpdateData;
import com.worldventures.dreamtrips.wallet.domain.entity.SmartCardFirmware;
import com.worldventures.dreamtrips.wallet.service.firmware.FirmwareRepository;

import javax.inject.Inject;
import javax.inject.Named;

import io.techery.janet.Command;
import io.techery.janet.Janet;
import rx.Observable;
import rx.Subscription;

import static com.worldventures.dreamtrips.core.janet.JanetModule.JANET_WALLET;

public abstract class BaseLoadFirmwareCommand extends Command<Void> implements InjectableAction {

   @Inject @Named(JANET_WALLET) Janet janet;
   @Inject FirmwareRepository firmwareRepository;

   @Override
   protected void run(CommandCallback<Void> callback) throws Throwable {
      callback.onProgress(0);

      Subscription subscription = provideProgress().subscribe(callback::onProgress);

      loadFile()
            .flatMap(aVoid -> updateFirmware())
            .subscribe(action -> {
               subscription.unsubscribe();
               callback.onSuccess(null);
            }, callback::onFail);
   }

   private Observable<Void> updateFirmware() {
      FirmwareUpdateData firmwareData = firmwareRepository.getFirmwareUpdateData();
      FirmwareUpdateData newFirmwareData = ImmutableFirmwareUpdateData.builder()
            .from(firmwareData)
            .currentFirmwareVersion(updatedSmartCardFirmware(firmwareData.currentFirmwareVersion()))
            .build();
      firmwareRepository.setFirmwareUpdateData(newFirmwareData);
      return Observable.just(null);
   }

   abstract Observable<Integer> provideProgress();

   abstract Observable<Void> loadFile();

   abstract SmartCardFirmware updatedSmartCardFirmware(SmartCardFirmware currentSmartCardFirmware);
}

File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/BaseLoadFirmwareCommand.kt
package com.worldventures.wallet.service.firmware.command

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.di.WalletJanetModule.JANET_WALLET
import com.worldventures.wallet.domain.entity.SmartCardFirmware
import com.worldventures.wallet.service.firmware.FirmwareRepository
import io.techery.janet.Command
import io.techery.janet.Janet
import rx.Observable
import javax.inject.Inject
import javax.inject.Named

abstract class BaseLoadFirmwareCommand : Command<Void>(), InjectableAction {

   @field:[Inject Named(JANET_WALLET)] lateinit var janet: Janet
   @Inject lateinit var firmwareRepository: FirmwareRepository

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<Void>) {
      val subscription = provideProgress().subscribe({ callback.onProgress(it) })

      loadFile()
            .flatMap { updateFirmware() }
            .subscribe({
               subscription.unsubscribe()
               callback.onSuccess(null)
            }, { callback.onFail(it) })
   }

   private fun updateFirmware(): Observable<Void> {
      val firmwareData = firmwareRepository.firmwareUpdateData
      val newFirmwareData = firmwareData.copy(currentFirmwareVersion = updatedSmartCardFirmware(firmwareData.currentFirmwareVersion))
      firmwareRepository.firmwareUpdateData = newFirmwareData
      return Observable.just(null)
   }

   internal abstract fun provideProgress(): Observable<Int>

   internal abstract fun loadFile(): Observable<Void>

   internal abstract fun updatedSmartCardFirmware(currentSmartCardFirmware: SmartCardFirmware): SmartCardFirmware
}
